### PR TITLE
Support builtin callable with object arguments in dynamo

### PR DESCRIPTION
**Pull request by laithsakka**
Fix issue #117556 
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118678

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by anijain2305 — APPROVED** [review]

File: test/dynamo/test_functions.py [hunk 1]
```diff
@@ -450,6 +450,31 @@ def test_callable_builtin(x):
...
+            return x + 1
+
+        input = torch.randn(4)
+
+        self.assertEqual(fn(input, NotCallableClass()), input)
+        self.assertEqual(fn(input, CallableClass()), input)
+
+        # passing tensor and scalars
+        self.assertEqual(fn(input, 1), input + 1)
+        self.assertEqual(fn(input, 1.1), input + 1)
```
Comment (anijain2305) [inline test/dynamo/test_functions.py:474]: Would suggest to use tensor here. There is some piece of code somewhere in Dynamo which check is this function is even worth of compilation, and it checks that either the inputs or the function itself should have something torch.Tensor related.

Comment (laithsakka) [reply test/dynamo/test_functions.py:474]: input is a tensor. I am not sure i understood the comment

**Comment by laithsakka** [comment]
1) update the unit test it sounds that dynamo does not get triggered when i pass scalars and tensor to 
     def fn1(x, arg):
            if callable(arg):
                return x
            return x + 1
           
           
    so I added fn2 to the test 
    2) added explicit handle to tensor and constant input and SymNodeVariable that they are not callable.

**Comment by laithsakka** [comment]
running  TORCH_LOGS="guards, graph"  python3 test/dynamo/test_functions.py -k  test_callable_class
gives
```
    ghstack checkout https://github.com/pytorch/pytorch/pull/118678

(base) [lsakka@devvm4561.ash0 /data/users/lsakka/pytorch (support_callable)]$ TORCH_LOGS="guards, graph"  python3 test/dynamo/test_functions.py -k  test_callable_class
[2024-01-30 16:08:31,329] [0/0] torch._dynamo.output_graph.__graph: [DEBUG] TRACED GRAPH
[2024-01-30 16:08:31,329] [0/0] torch._dynamo.output_graph.__graph: [DEBUG]  __compiled_fn_0 /data/users/lsakka/pytorch/torch/fx/_lazy_graph_module.py opcode         name    target                   args       kwargs
[2024-01-30 16:08:31,329] [0/0] torch._dynamo.output_graph.__graph: [DEBUG] -------------  ------  -----------------------  ---------  --------
[2024-01-30 16:08:31,329] [0/0] torch._dynamo.output_graph.__graph: [DEBUG] placeholder    l_x_    L_x_                     ()         {}
[2024-01-30 16:08:31,329] [0/0] torch._dynamo.output_graph.__graph: [DEBUG] call_function  add     <built-in function add>  (l_x_, 1)  {}
[2024-01-30 16:08:31,329] [0/0] torch._dynamo.output_graph.__graph: [DEBUG] output         output  output                   ((add,),)  {}
[2024-01-30 16:08:31,329] [0/0] torch._dynamo.output_graph.__graph: [DEBUG] 
[2024-01-30 16:08:31,345] [0/0] torch._dynamo.guards.__guards: [DEBUG] GUARDS:
[2024-01-30 16:08:31,345] [0/0] torch._dynamo.guards.__guards: [DEBUG] hasattr(L['x'], '_dynamo_dynamic_indices') == False           # return x + 1  # test/dynamo/test_functions.py:465 in fn1
[2024-01-30 16:08:31,345] [0/0] torch._dynamo.guards.__guards: [DEBUG] ___check_type_id(L['arg'], 317807040)                         # if callable(arg):  # test/dynamo/test_functions.py:463 in fn1
[2024-01-30 16:08:31,346] [0/0] torch._dynamo.guards.__guards: [DEBUG] utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:388 in init_ambient_guards
[2024-01-30 16:08:31,346] [0/0] torch._dynamo.guards.__guards: [DEBUG] ___check_current_backend(140059508756240)                     # _dynamo/output_graph.py:394 in init_ambient_guards
[2024-01-30 16:08:31,346] [0/0] torch._dynamo.guards.__guards: [DEBUG] check_tensor(L['x'], Tensor, DispatchKeySet(CPU, BackendSelect, ADInplaceOrView, AutogradCPU), torch.float32, device=None, requires_grad=False, size=[4], stride=[1])  # return x + 1  # test/dynamo/test_functions.py:465 in fn1
[2024-01-30 16:08:31,376] [1/0] torch._dynamo.output_graph.__graph: [DEBUG] TRACED GRAPH
[2024-01-30 16:08:31,376] [1/0] torch._dynamo.output_graph.__graph: [DEBUG]  __compiled_fn_1 /data/users/lsakka/pytorch/torch/fx/_lazy_graph_module.py opcode         name    target                   args       kwargs
[2024-01-30 16:08:31,376] [1/0] torch._dynamo.output_graph.__graph: [DEBUG] -------------  ------  -----------------------  ---------  --------
[2024-01-30 16:08:31,376] [1/0] torch._dynamo.output_graph.__graph: [DEBUG] placeholder    l_x_    L_x_                     ()         {}
[2024-01-30 16:08:31,376] [1/0] torch._dynamo.output_graph.__graph: [DEBUG] call_function  add     <built-in function add>  (l_x_, 1)  {}
[2024-01-30 16:08:31,376] [1/0] torch._dynamo.output_graph.__graph: [DEBUG] output         output  output                   ((add,),)  {}
[2024-01-30 16:08:31,376] [1/0] torch._dynamo.output_graph.__graph: [DEBUG] 
[2024-01-30 16:08:31,380] [1/0] torch._dynamo.guards.__guards: [DEBUG] GUARDS:
[2024-01-30 16:08:31,380] [1/0] torch._dynamo.guards.__guards: [DEBUG] hasattr(L['x'], '_dynamo_dynamic_indices') == False           # return x + 1  # test/dynamo/test_functions.py:471 in fn2
[2024-01-30 16:08:31,380] [1/0] torch._dynamo.guards.__guards: [DEBUG] ___check_type_id(L['arg'], 317807040)                         # if callable(arg):  # test/dynamo/test_functions.py:469 in fn2
[2024-01-30 16:08:31,380] [1/0] torch._dynamo.guards.__guards: [DEBUG] utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:388 in init_ambient_guards
[2024-01-30 16:08:31,381] [1/0] torch._dynamo.guards.__guards: [DEBUG] ___check_current_backend(140059508930000)                     # _dynamo/output_graph.py:394 in init_ambient_guards
[2024-01-30 16:08:31,381] [1/0] torch._dynamo.guards.__guards: [DEBUG] check_tensor(L['x'], Tensor, DispatchKeySet(CPU, BackendSelect, ADInplaceOrView, AutogradCPU), torch.float32, device=None, requires_grad=False, size=[4], stride=[1])  # return x + 1  # test/dynamo/test_functions.py:471 in fn2
[2024-01-30 16:08:31,392] [1/1] torch._dynamo.output_graph.__graph: [DEBUG] TRACED GRAPH
[2024-01-30 16:08:31,392] [1/1] torch._dynamo.output_graph.__graph: [DEBUG]  __compiled_fn_2 /data/users/lsakka/pytorch/torch/fx/_lazy_graph_module.py opcode         name    target                   args       kwargs
[2024-01-30 16:08:31,392] [1/1] torch._dynamo.output_graph.__graph: [DEBUG] -------------  ------  -----------------------  ---------  --------
[2024-01-30 16:08:31,392] [1/1] torch._dynamo.output_graph.__graph: [DEBUG] placeholder    l_x_    L_x_                     ()         {}
[2024-01-30 16:08:31,392] [1/1] torch._dynamo.output_graph.__graph: [DEBUG] call_function  mul     <built-in function mul>  (l_x_, 2)  {}
[2024-01-30 16:08:31,392] [1/1] torch._dynamo.output_graph.__graph: [DEBUG] output         output  output                   ((mul,),)  {}
[2024-01-30 16:08:31,392] [1/1] torch._dynamo.output_graph.__graph: [DEBUG] 
[2024-01-30 16:08:31,395] [1/1] torch._dynamo.guards.__guards: [DEBUG] GUARDS:
[2024-01-30 16:08:31,396] [1/1] torch._dynamo.guards.__guards: [DEBUG] hasattr(L['x'], '_dynamo_dynamic_indices') == False           # return x * 2  # test/dynamo/test_functions.py:470 in fn2
[2024-01-30 16:08:31,396] [1/1] torch._dynamo.guards.__guards: [DEBUG] ___check_type_id(L['arg'], 317720192)                         # if callable(arg):  # test/dynamo/test_functions.py:469 in fn2
[2024-01-30 16:08:31,396] [1/1] torch._dynamo.guards.__guards: [DEBUG] utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:388 in init_ambient_guards
[2024-01-30 16:08:31,396] [1/1] torch._dynamo.guards.__guards: [DEBUG] ___check_current_backend(140059508930000)                     # _dynamo/output_graph.py:394 in init_ambient_guards
[2024-01-30 16:08:31,396] [1/1] torch._dynamo.guards.__guards: [DEBUG] check_tensor(L['x'], Tensor, DispatchKeySet(CPU, BackendSelect, ADInplaceOrView, AutogradCPU), torch.float32, device=None, requires_grad=False, size=[4], stride=[1])  # return x * 2  # test/dynamo/test_functions.py:470 in fn2
[2024-01-30 16:08:31,406] [1/2] torch._dynamo.output_graph.__graph: [DEBUG] TRACED GRAPH
[2024-01-30 16:08:31,406] [1/2] torch._dynamo.output_graph.__graph: [DEBUG]  __compiled_fn_3 /data/users/lsakka/pytorch/torch/fx/_lazy_graph_module.py opcode         name    target                   args       kwargs
[2024-01-30 16:08:31,406] [1/2] torch._dynamo.output_graph.__graph: [DEBUG] -------------  ------  -----------------------  ---------  --------
[2024-01-30 16:08:31,406] [1/2] torch._dynamo.output_graph.__graph: [DEBUG] placeholder    l_x_    L_x_                     ()         {}
[2024-01-30 16:08:31,406] [1/2] torch._dynamo.output_graph.__graph: [DEBUG] call_function  add     <built-in function add>  (l_x_, 1)  {}
[2024-01-30 16:08:31,406] [1/2] torch._dynamo.output_graph.__graph: [DEBUG] output         output  output                   ((add,),)  {}
[2024-01-30 16:08:31,406] [1/2] torch._dynamo.output_graph.__graph: [DEBUG] 
[2024-01-30 16:08:31,409] [1/2] torch._dynamo.guards.__guards: [DEBUG] GUARDS:
[2024-01-30 16:08:31,409] [1/2] torch._dynamo.guards.__guards: [DEBUG] hasattr(L['x'], '_dynamo_dynamic_indices') == False           # return x + 1  # test/dynamo/test_functions.py:471 in fn2
[2024-01-30 16:08:31,409] [1/2] torch._dynamo.guards.__guards: [DEBUG] ___check_type_id(L['arg'], 8815104)                           # if callable(arg):  # test/dynamo/test_functions.py:469 in fn2
[2024-01-30 16:08:31,409] [1/2] torch._dynamo.guards.__guards: [DEBUG] L['arg'] == 1                                                 # if callable(arg):  # test/dynamo/test_functions.py:469 in fn2
[2024-01-30 16:08:31,410] [1/2] torch._dynamo.guards.__guards: [DEBUG] utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:388 in init_ambient_guards
[2024-01-30 16:08:31,410] [1/2] torch._dynamo.guards.__guards: [DEBUG] ___check_current_backend(140059508930000)                     # _dynamo/output_graph.py:394 in init_ambient_guards
[2024-01-30 16:08:31,410] [1/2] torch._dynamo.guards.__guards: [DEBUG] check_tensor(L['x'], Tensor, DispatchKeySet(CPU, BackendSelect, ADInplaceOrView, AutogradCPU), torch.float32, device=None, requires_grad=False, size=[4], stride=[1])  # return x + 1  # test/dynamo/test_functions.py:471 in fn2
[2024-01-30 16:08:31,418] [1/3] torch._dynamo.output_graph.__graph: [DEBUG] TRACED GRAPH
[2024-01-30 16:08:31,418] [1/3] torch._dynamo.output_graph.__graph: [DEBUG]  __compiled_fn_4 /data/users/lsakka/pytorch/torch/fx/_lazy_graph_module.py opcode         name    target                   args       kwargs
[2024-01-30 16:08:31,418] [1/3] torch._dynamo.output_graph.__graph: [DEBUG] -------------  ------  -----------------------  ---------  --------
[2024-01-30 16:08:31,418] [1/3] torch._dynamo.output_graph.__graph: [DEBUG] placeholder    l_x_    L_x_                     ()         {}
[2024-01-30 16:08:31,418] [1/3] torch._dynamo.output_graph.__graph: [DEBUG] call_function  add     <built-in function add>  (l_x_, 1)  {}
[2024-01-30 16:08:31,418] [1/3] torch._dynamo.output_graph.__graph: [DEBUG] output         output  output                   ((add,),)  {}
[2024-01-30 16:08:31,418] [1/3] torch._dynamo.output_graph.__graph: [DEBUG] 
[2024-01-30 16:08:31,421] [1/3] torch._dynamo.guards.__guards: [DEBUG] GUARDS:
[2024-01-30 16:08:31,421] [1/3] torch._dynamo.guards.__guards: [DEBUG] hasattr(L['x'], '_dynamo_dynamic_indices') == False           # return x + 1  # test/dynamo/test_functions.py:471 in fn2
[2024-01-30 16:08:31,421] [1/3] torch._dynamo.guards.__guards: [DEBUG] ___check_type_id(L['arg'], 8824544)                           # if callable(arg):  # test/dynamo/test_functions.py:469 in fn2
[2024-01-30 16:08:31,422] [1/3] torch._dynamo.guards.__guards: [DEBUG] L['arg'] == 1.1                                               # if callable(arg):  # test/dynamo/test_functions.py:469 in fn2
[2024-01-30 16:08:31,422] [1/3] torch._dynamo.guards.__guards: [DEBUG] utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:388 in init_ambient_guards
[2024-01-30 16:08:31,422] [1/3] torch._dynamo.guards.__guards: [DEBUG] ___check_current_backend(140059508930000)                     # _dynamo/output_graph.py:394 in init_ambient_guards
[2024-01-30 16:08:31,422] [1/3] torch._dynamo.guards.__guards: [DEBUG] check_tensor(L['x'], Tensor, DispatchKeySet(CPU, BackendSelect, ADInplaceOrView, AutogradCPU), torch.float32, device=None, requires_grad=False, size=[4], stride=[1])  # return x + 1  # test/dynamo/test_functions.py:471 in fn2
[2024-01-30 16:08:31,430] [1/4] torch._dynamo.output_graph.__graph: [DEBUG] TRACED GRAPH
[2024-01-30 16:08:31,430] [1/4] torch._dynamo.output_graph.__graph: [DEBUG]  __compiled_fn_5 /data/users/lsakka/pytorch/torch/fx/_lazy_graph_module.py opcode         name    target                   args       kwargs
[2024-01-30 16:08:31,430] [1/4] torch._dynamo.output_graph.__graph: [DEBUG] -------------  ------  -----------------------  ---------  --------
[2024-01-30 16:08:31,430] [1/4] torch._dynamo.output_graph.__graph: [DEBUG] placeholder    l_x_    L_x_                     ()         {}
[2024-01-30 16:08:31,430] [1/4] torch._dynamo.output_graph.__graph: [DEBUG] call_function  add     <built-in function add>  (l_x_, 1)  {}
[2024-01-30 16:08:31,430] [1/4] torch._dynamo.output_graph.__graph: [DEBUG] output         output  output                   ((add,),)  {}
[2024-01-30 16:08:31,430] [1/4] torch._dynamo.output_graph.__graph: [DEBUG] 
[2024-01-30 16:08:31,433] [1/4] torch._dynamo.guards.__guards: [DEBUG] GUARDS:
[2024-01-30 16:08:31,433] [1/4] torch._dynamo.guards.__guards: [DEBUG] hasattr(L['x'], '_dynamo_dynamic_indices') == False           # return x + 1  # test/dynamo/test_functions.py:471 in fn2
[2024-01-30 16:08:31,433] [1/4] torch._dynamo.guards.__guards: [DEBUG] ___check_obj_id(L['arg'], 8883552)                            # if callable(arg):  # test/dynamo/test_functions.py:469 in fn2
[2024-01-30 16:08:31,433] [1/4] torch._dynamo.guards.__guards: [DEBUG] utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:388 in init_ambient_guards
[2024-01-30 16:08:31,433] [1/4] torch._dynamo.guards.__guards: [DEBUG] ___check_current_backend(140059508930000)                     # _dynamo/output_graph.py:394 in init_ambient_guards
[2024-01-30 16:08:31,433] [1/4] torch._dynamo.guards.__guards: [DEBUG] check_tensor(L['x'], Tensor, DispatchKeySet(CPU, BackendSelect, ADInplaceOrView, AutogradCPU), torch.float32, device=None, requires_grad=False, size=[4], stride=[1])  # return x + 1  # test/dynamo/test_functions.py:471 in fn2
[2024-01-30 16:08:31,441] [1/5] torch._dynamo.output_graph.__graph: [DEBUG] TRACED GRAPH
[2024-01-30 16:08:31,441] [1/5] torch._dynamo.output_graph.__graph: [DEBUG]  __compiled_fn_6 /data/users/lsakka/pytorch/torch/fx/_lazy_graph_module.py opcode         name    target                   args         kwargs
[2024-01-30 16:08:31,441] [1/5] torch._dynamo.output_graph.__graph: [DEBUG] -------------  ------  -----------------------  -----------  --------
[2024-01-30 16:08:31,441] [1/5] torch._dynamo.output_graph.__graph: [DEBUG] placeholder    l_arg_  L_arg_                   ()           {}
[2024-01-30 16:08:31,441] [1/5] torch._dynamo.output_graph.__graph: [DEBUG] call_function  add     <built-in function add>  (l_arg_, 1)  {}
[2024-01-30 16:08:31,441] [1/5] torch._dynamo.output_graph.__graph: [DEBUG] output         output  output                   ((add,),)    {}
[2024-01-30 16:08:31,441] [1/5] torch._dynamo.output_graph.__graph: [DEBUG] 
[2024-01-30 16:08:31,444] [1/5] torch._dynamo.guards.__guards: [DEBUG] GUARDS:
[2024-01-30 16:08:31,444] [1/5] torch._dynamo.guards.__guards: [DEBUG] L['arg'] is L['x']                                            # return x + 1  # test/dynamo/test_functions.py:471 in fn2
[2024-01-30 16:08:31,444] [1/5] torch._dynamo.guards.__guards: [DEBUG] hasattr(L['arg'], '_dynamo_dynamic_indices') == False         # if callable(arg):  # test/dynamo/test_functions.py:469 in fn2
[2024-01-30 16:08:31,444] [1/5] torch._dynamo.guards.__guards: [DEBUG] utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:388 in init_ambient_guards
[2024-01-30 16:08:31,445] [1/5] torch._dynamo.guards.__guards: [DEBUG] ___check_current_backend(140059508930000)                     # _dynamo/output_graph.py:394 in init_ambient_guards
[2024-01-30 16:08:31,445] [1/5] torch._dynamo.guards.__guards: [DEBUG] check_tensor(L['arg'], Tensor, DispatchKeySet(CPU, BackendSelect, ADInplaceOrView, AutogradCPU), torch.float32, device=None, requires_grad=False, size=[4], stride=[1])  # if callable(arg):  # test/dynamo/test_functions.py:469 in fn2
stats [('calls_captured', 7), ('unique_graphs', 7)]
.
```

**Pull request closed without merging**